Dockerise

mattcen opened the pull request:
Hi @cassar,

I wanted to make it easier to run a local instance of Hackerspace for dev/testing, and I've been playing with Docker. This PR makes that happen, but I'm unhappy with the changes I've had to make to `config/database.yml` to make the database config work. I don't understand enough about how Rails environments work to know if there's a good way to have a config that works for Docker but doesn't conflict with non-Docker instances.

1. Do you have any feedback on how I can better make DB config work?
2. You had commented out `mini_racer` in `Gemfile`, and without it, I couldn't get Hackerspace working because it didn't have a JavaScript interpreter. Is there a good reason for having this commented out? Is there another JS library I should be using? 

This is far from polished, but to get this running (assuming you have Docker and docker-compose installed and configured), all you need to do is:

```sh
git clone -b dockerise https://github.com/mattcen/hackerspace3
cd hackerspace3
docker pull mattcen/hackerspace3 # If you don't want to wait for the image to build locally
./docker-setup.sh
docker-compose up
```

mattcen commented:
I also intend to find a more secure way to specify a password, rather than store it in `docker-compose.yml`.

cassar commented:
Hey @mattcen, 

Really great you're doing this, setting up environments is really unpleasant.

> 1. Do you have any feedback on how I can better make DB config work?

Take a look at Mastodon's `database.yml` [here](https://github.com/tootsuite/mastodon/blob/master/config/database.yml).
it looks pretty flexible. 

And its [dockfile](https://github.com/tootsuite/mastodon/blob/master/Dockerfile) would also be a good reference.

> 2. You had commented out mini_racer in Gemfile, and without it

These are included (commented out) by default in new rails apps as hints, all I did was move them to the bottom (out of the way) in the Gemfile. 

Go ahead and turn on whatever you need.

> I also intend to find a more secure way to specify a password,

If it's just used for development and testing environment's I wouldn't worry about it (yet)

mattcen commented:
OK, @cassar, I reckon this should be good to go now. Can you just check that the changes to `database.yml` don't make it too difficult for you to deploy without Docker? If that's good, then I'll merge it!

yumiris commented:
In regard to securely storing passwords rather than exposing them in `docker-compose.yml`, referencing an environment file should do the trick:

```yaml
services
  hackerspace3:
    container_name: hackerspace3
    image: hackerspace3
    env_file:
     - ./.env
```

As for the .env file's contents:
```sh
POSTGRES_PASSWORD: SuperSecretPass
POSTGRES_USER: hackerspace
POSTGRES_HOST: postgres
POSTGRES_DB: hackerspace3

# etc...
```

For securely its structure it in the repository:

```sh
cat .env > .env.example && echo ".env" >> .gitignore

# edit .env.example with example values instead of the secure ones

git add .env.example && git commit -m "Added example .env file; use cp .env.example .env for the magic!"
```

mattcen commented:
Thanks @yumiris, good plan; have updated the PR (took me a while to debug why your `.env` file example wasn't working: needed to replace `: ` with `=`!)

yumiris commented:
Bah, my bad. One would think that `docker-compose.yml` permitting `:` and `=` interchangeably would be the same for `.env` files. Hopefully it wasn't too much of a headache to get it fixed :p

mattcen commented:
Haha, all good; I guess the reason they require `key=value` format for the `env_file` contents is because then it  can also be sourced by the shell and the syntax will work OK.
I ended up reading the Docker documentation but not noticing the typo, and eventually found a bug where someone had the same issue as me. :)

cassar commented on Gemfile.lock:
```diff
@@ -169,6 +169,7 @@ GEM
       addressable (~> 2.7)
     letter_opener (1.7.0)
       launchy (~> 2.2)
+    libv8 (7.3.492.27.1)
```
Once you do a `bundle install` this dependency tree gets updated and also needs to be committed.

mattcen replied:
That makes sense; thanks!

cassar commented on config/database.yml:
```diff
@@ -29,7 +29,7 @@ development:
   # To create additional roles in postgres see `$ createuser --help`.
   # When left blank, postgres will use the default role. This is
   # the same name as the operating system user that initialized the database.
-  username: <%= ENV['POSTGRES_USER'] || 'hackerspace' %>
```
Just reverting these to default to empty values for development and test environments if that's ok.

mattcen replied:
Ah yeah, sure; it wasn't clear to me what the default values should've been :)

cassar commented on Gemfile:
```diff
@@ -53,6 +53,8 @@ gem 'omniauth-rails_csrf_protection'
 gem 'webpacker'
 # A tagging plugin that allows for custom tagging along dynamic contexts.
 gem 'acts-as-taggable-on', '~> 6.0'
+# See https://github.com/rails/execjs#readme for more supported runtimes
```
Have just moved this out of the optional extras group into our core group of gems.

mattcen replied:
Cool. :)

mattcen commented:
Sweet, thanks for all that, @cassar! I've squashed those commits now. (Incidentally, I did *not* know that you'd be able to make changes to a branch on *my* fork of the repo! That's both handy and a little concerning! 😄  )

cassar commented:
Ooo yeah, I suppose all bets are off when you put a PR up.

I checked out the PR branch and pushed to that. 🤷‍♂️

mattcen commented:
Huh, interesting! I forgot you could check out a PR, too! In any case, all good!

waiholiu commented:
Hi,

Is this PR intended just for deployment purposes? Or to set up an environment for development purposes? Cause it was super easy to set up and running but how do I make changes to my solution and have it show up in my local instance?

I haven't played around with docker much but docker-compose file looks like it's just getting the image from mattcen/hackerspace

mattcen commented:
Hi @waiholiu!

This PR could be used for both dev/test *and* production without much effort. I've just updated the Docker section of [README.md](https://github.com/govhackaustralia/hackerspace3/blob/main/README.md) with additional information that should answer your question. Could you please take a look at that and see if it makes sense?

To answer your question about getting the image from `mattcen/hackerspace3`, (1) I've just now updated this to pull it from `govhackau/hackerspace3`, now the PR has been merged, and (2), by default, `docker-compose` will build the image from scratch if it doesn't find a local copy. This can be avoided by manually pulling the image, which should save a couple of minutes of local compute time. This is also now made more clear in the [README.md](https://github.com/govhackaustralia/hackerspace3/blob/main/README.md).

waiholiu commented:
@mattcen nice one! 

I can get it working with those commands.

I did have to run the below commands each time I make a change. 

docker build -t govhackau/hackerspace3 .
docker-compose run --rm -v "$PWD":/usr/src/app --service-ports hackerspace3

or 

docker-compose up --build



Which doesn't take long but it does seem like i'm building the image each time so not sure if the docker-compose run --rm -v "$PWD":/usr/src/app --service-ports hackerspace3 is working properly

mattcen commented:
@waiholiu You definitely shouldn't need to rebuild the image every time you make a change (which is what those two sets of commands you've mentioned do). I've just made a small tweak to the README again, to add the `-d` option to ` docker-compose run --rm -d -v "$PWD":/usr/src/app --service-ports hackerspace3`, so that it backgrounds the container instead of leave it in the foreground (which shouldn't make a difference to the behaviour you were seeing anyway).

I recommend trying this (Note that these commands *all* need to be run at the top level of the repository directory tree - the same directory that `docker-compose.yml` lives in):

```bash
docker-compose down # Just to ensure we don't have existing containers running
docker-compose up -d postgres
docker-compose run --name testHackerspace --rm -d -v "$PWD":/usr/src/app --service-ports hackerspace3
touch ThisIsATestFile.txt
docker exec -it testHackerspace ls -l ThisIsATestFile.txt
# Or
docker exec -it testHackerspace bash
```

These last two commands will run `ls -l ThisIsATestFile.txt` and `bash` respectively in the container just started with `docker-compose run`, and the latter will let you look around inside it. This should confirm that the file you just created (`ThisIsATestFile.txt`) is visible in the container.

In theory, when you change a file in use by Rails, Rails will rebuild itself to use that new file on the fly.

It's also worth being aware of `docker logs -f ContainerName` (where `ContainerName` is `testHackerspace` in the above example). This will show you the logs of the running container as it renders pages.

Before I edited `app/views/competitions/show.html.haml` to change the "Welcome to GovHack" line, and when I refreshed the page, the changed text appeared immediately.

If none of this works for you, perhaps an instant message chat will help sort it out. I'm available on the GovHackHQ slack team if you want to ping me there.

waiholiu commented:
Hi @mattcen ,

i tried what you wrote, can confirm the testisatestfile.txt isn't actually copied in the container until I rebuild the image.

Funnily enough, I set it up on my mac and this works as you described. I'm guessing this is a windows docker issue maybe? Might ask on the slack chat whether people on windows are experiencing the same issue.

but I think it's probably fine - I just chucked this into a bash script

docker-compose rm -f -s -v hackerspace3 
docker build -t govhackau/hackerspace3 . 
docker-compose run --rm -d -v "$PWD":/usr/src/app --service-ports hackerspace3

and run it whenever i wanna refresh the site. It does it in a few seconds anyway.

mattcen commented:
@waiholiu I'll respond in more detail on Slack, but "$PWD" probably won't work on Windows; I suspect that's your problem.
